### ISSUE TITLE
WASM support for training

### ISSUE THREAD

**Issue by nathanielsimard** [tracel-ai/burn#1254]
Support training on the wasm target. For this, we probably need to have alternative implementations of file checkpointers, file loggers, and similar components. It's not clear whether we want to support training in the browser, or if supporting wasm runtimes alone is sufficient. In the latter case, we could use an alternative file system API provided by such a runtime.

**Requires** #1253

**Comment by antimora** [comment]
I am personally not in favor we support training in the browser. This is such niche use case that requires the code base to accommodate.

**Comment by Luni-4** [comment]
In my opinion we can support training in the browser at a later time in a new crate called `burn-wasm`. In this way we do not have to change the other `burn` crates, but we can create an independent piece of software with the goal of R&D about the potentiality of training a neural network inside a browser. So for me it is a yes, but with the label of R&D

**Comment by nathanielsimard** [comment]
I think we can actually support training in the browser without having to change much of the architecture. It's similar to how we can support `no-std`: by having primitive type stubs in `burn-common`. So when #1250 is done, it wouldn't be that hard to support training on the web.

We have to keep in mind that supporting `wasm` can be beneficial as a deployment format, which can be used similarly to Docker, so it can be useful.

**Comment by ArthurBrussee** [comment]
Also just to chime in - I imagine training massive LLMs or even big convnets in the browsers is niche. However, consider NeRF and gaussian splat like models, training in a browser could be totally fine & useful. Generally, being able to use ML as "just" good numerical optimization is nice! It's not all about massive models.

**Comment by JulienVig** [comment]
Hi, our team would be interested in migrating our project to Burn if web training was supported.
We're implementing a [web platform](https://github.com/epfml/disco/tree/develop) for federated learning and we rely on web training to lower the barrier-to-entry for non-technical people to participate in a training session (no need to install any software or code anything if everything happens in the browser). However, the main web training frameworks, Tensorflow.js and ONNX Runtime Web are [deprecated](https://github.com/microsoft/onnxruntime/issues/22799#issuecomment-2468982759) or slowly going unmaintained, so we're actively looking for alternatives and Burn seems like a promising one.

Do you foresee web training support coming any time soon? If not, could we contribute to adding web training support to Burn in some way?

**Comment by nathanielsimard** [comment]
@JulienVig we already support training on the web, but you have to write your own training loop, since most of our utilities assume a file system to log stuff into (checkpointer, metrics, etc.)
